stdlib/mexpr/type-check.mc

```diff
@@ -0,0 +1,436 @@
+----------------------
+
+lang Unify = MExprEq
+  -- Unify the types `ty1' and `ty2'. Modifies the types in place.
+  sem unify =
+  | (ty1, ty2) -> unifyWithNames biEmpty (ty1, ty2)
+
+  sem unifyWithNames (names : BiNameMap) =
+  | (ty1, ty2) ->
+    -- OPT(aathn, 2021-09-27): This equality check traverses the types unnecessarily.
+    -- TODO(aathn, 2021-09-28): This equality check uses empty type environment.
+    if eqType ty1 ty2 then
+      ()
+    else
+      unifyBase names (ty1, ty2)
+
+  sem unifyBase (names : BiNameMap) =
+  | (ty1, ty2) ->
+    unificationError (ty1, ty2)
+
+  sem unificationError =
+  | (ty1, ty2) ->
+    let msg = join [
+      "Type check failed: unification failure\n",
+      "LHS: ", _pprintType ty1, "\n",
+      "RHS: ", _pprintType ty2, "\n"
+    ] in
+    infoErrorExit (infoTy ty1) msg
+
+  sem occurs (tv : TVar) =
+  -- Intentionally left empty
+end
+
+lang VarTypeUnify = Unify + VarTypeAst + UnknownTypeAst
+  sem unifyBase (names : BiNameMap) =
+  -- We don't unify variables with TyUnknown
+  | (TyFlex {contents = r}, !TyUnknown _ & ty1)
+  | (!TyUnknown _ & ty1, TyFlex {contents = r}) ->
+    match deref r with Unbound _ & tv then
+      occurs tv ty1; modref r (Link ty1)
+    else match deref r with Link ty2 then
+      unifyWithNames names (ty1, ty2)
+    else never
+  | (TyVar t1 & ty1, TyVar t2 & ty2) ->
+    if not (nameEq t1.ident t2.ident) then
+      match biLookup (t1.ident, t2.ident) names with None () then
+        unificationError (ty1, ty2)
+      else ()
+    else ()
+
+  sem occurs (tv : TVar) =
+  | TyFlex {info = info, contents = r} ->
+    -- TODO(aathn, 2021-09-28): This equality check uses empty type environment.
+    if eqTVar [] (deref r, tv) then
+      let msg = "Type check failed: occurs check\n" in
+      infoErrorExit info msg
+    else
+      match deref r with Link ty then
+        occurs tv ty
+      else match deref r with Unbound ({level = k} & t) then
+        let minLevel =
+          match tv with Unbound {level = l} then mini k l else k
+        in
+        modref r (Unbound {t with level = minLevel})
+      else never
+  | TyVar _ ->
+    ()
+end
+
+lang FunTypeUnify = Unify + FunTypeAst
+  sem unifyBase (names : BiNameMap) =
+  | (TyArrow {from = from1, to = to1}, TyArrow {from = from2, to = to2}) ->
+    unifyWithNames names (from1, from2);
+    unifyWithNames names (to1, to2)
+
+  sem occurs (tv : TVar) =
+  | TyArrow {from = from, to = to} ->
+    occurs tv from;
+    occurs tv to
+end
+
+lang AllTypeUnify = Unify + AllTypeAst
+  sem unifyBase (names : BiNameMap) =
+  | (TyAll t1, TyAll t2) ->
+    unifyWithNames (biInsert (t1.ident, t2.ident) names) (t1.ty, t2.ty)
+
+  sem occurs (tv : TVar) =
+  | TyAll t ->
+    match tv with Unbound {weak = true} then
+      let msg = join [
+        "Type check failed: unification failure\n",
+        "Attempted to unify monomorphic type variable with polymorphic type!\n"
+      ] in
+      infoErrorExit t.info msg
```

Comment:
`occurs` currently performs three tasks, all of which require traversing the type to unify with, in a single traversal.
- Performing an occurs check
- Updating `level` fields of type variables
- Checking that monomorphic type variables are not unified with polymorphic types

I should probably rename the function, and add a comment explaining this.

It would not be possible to unify `_a` with `[all b. b->b]`, correct. The way I interpreted the paper, this is to prevent phenomena like the following,  where

    bad1 = lam f. (poly `f, (f 42) + 1)

type checks, but

    bad2 = lam f. ((f 42) + 1, poly `f)

does not (with `poly : (all a. a -> a) -> (Int, Bool)`).

Reply:
But the issue in that example is about unifying the type of `f` with `all a. a -> a` vs `Int -> Int`, but `[all b. b -> b]` is guarded by the `[]` constructor (the difference between full monotypes and guarded types in the paper). I feel like that difference should matter, shouldn't it?